### DISCUSSION TITLE
Models not generating .npz files after training in colab

### DISCUSSION THREAD

**Discussion by tacytown** [talmolab/sleap#1374] [Help!] [answered]
Hello! I'm training single animal models in colab and want to evaluate them like in the [Model evaluation](https://sleap.ai/notebooks/Model_evaluation.html) notebook, but my models folder isn't generating .npz files. I verified that training_config.json has outputs set to true (as in discussion [#1286](https://github.com/talmolab/sleap/discussions/1286)). A minority of trials did generate the .npz files, but I can't figure out anything that was done differently in those cases, or how to replicate them. 

I've included screenshots from google drive of a model that has the .npz files and one that doesn't.

<img width="317" alt="sleap-present" src="https://github.com/talmolab/sleap/assets/97413357/f2982f08-8a4c-4ff6-85bd-88bfed6dac71">
<img width="223" alt="sleap-missing" src="https://github.com/talmolab/sleap/assets/97413357/f3d01ed8-eb66-4d7e-b076-8692e57a2207">

Thanks so much for any help or suggestions!

**Comment by talmo** [comment]
Hi @tacytown,

Have you taken a look at the model results or the `viz` folder? This typically happens when training doesn't work very well and SLEAP fails to get any predictions from which to compute the metrics.

Can you paste the contents of `training_config.json` and `training_log.csv`? These can help us diagnose whether training completed successfully or if there's another issue.

Thanks!

Talmo

**Comment by tacytown** [comment]
Hello @talmo,

Thank you so much! The model results look great when I view the predictions.slp file in SLEAP, but in the viz folder the images appear slightly offset:
<img width="559" alt="validation 0024" src="https://github.com/talmolab/sleap/assets/97413357/284a7b22-cb07-4d2c-ae02-68c1feefb2d2">

Here's the contents of training_config.json:
```
{
    "data": {
        "labels": {
            "training_labels": "labels.v001.pkg.slp",
            "validation_labels": null,
            "validation_fraction": 0.1,
            "test_labels": null,
            "split_by_inds": false,
            "training_inds": [
                48,
                75,
                97,
                8,
                32,
                66,
                64,
                7,
                73,
                18,
                14,
                13,
                41,
                82,
                69,
                95,
                1,
                62,
                27,
                17,
                15,
                25,
                68,
                23,
                55,
                10,
                96,
                80,
                44,
                43,
                85,
                93,
                9,
                26,
                83,
                47,
                33,
                39,
                28,
                38,
                3,
                87,
                84,
                20,
                52,
                12,
                86,
                94,
                81,
                50,
                36,
                76,
                21,
                6,
                71,
                54,
                40,
                51,
                30,
                89,
                61,
                90,
                16,
                34,
                79,
                60,
                67,
                49,
                4,
                74,
                56,
                58,
                88,
                92,
                72,
                63,
                42,
                22,
                70,
                46,
                77,
                98,
                53,
                35,
                29,
                24,
                0,
                37,
                2,
                19
            ],
            "validation_inds": [
                59,
                31,
                5,
                57,
                99,
                45,
                91,
                11,
                65,
                78
            ],
            "test_inds": null,
            "search_path_hints": [
                ""
            ],
            "skeletons": [
                {
                    "directed": true,
                    "graph": {
                        "name": "Skeleton-85",
                        "num_edges_inserted": 15
                    },
                    "links": [
                        {
                            "edge_insert_idx": 0,
                            "key": 0,
                            "source": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "L1",
                                        1.0
                                    ]
                                }
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "L2",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/reduce": [
                                    {
                                        "py/type": "sleap.skeleton.EdgeType"
                                    },
                                    {
                                        "py/tuple": [
                                            1
                                        ]
                                    }
                                ]
                            }
                        },
                        {
                            "edge_insert_idx": 4,
                            "key": 0,
                            "source": {
                                "py/id": 2
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "L3",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 5,
                            "key": 0,
                            "source": {
                                "py/id": 2
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "L4",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 6,
                            "key": 0,
                            "source": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "R1",
                                        1.0
                                    ]
                                }
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "R2",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 7,
                            "key": 0,
                            "source": {
                                "py/id": 7
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "R3",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 8,
                            "key": 0,
                            "source": {
                                "py/id": 7
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "R4",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 9,
                            "key": 0,
                            "source": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "Head",
                                        1.0
                                    ]
                                }
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "Middle",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 10,
                            "key": 0,
                            "source": {
                                "py/id": 11
                            },
                            "target": {
                                "py/object": "sleap.skeleton.Node",
                                "py/state": {
                                    "py/tuple": [
                                        "End",
                                        1.0
                                    ]
                                }
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 13,
                            "key": 0,
                            "source": {
                                "py/id": 11
                            },
                            "target": {
                                "py/id": 1
                            },
                            "type": {
                                "py/id": 3
                            }
                        },
                        {
                            "edge_insert_idx": 14,
                            "key": 0,
                            "source": {
                                "py/id": 11
                            },
                            "target": {
                                "py/id": 6
                            },
                            "type": {
                                "py/id": 3
                            }
                        }
                    ],
                    "multigraph": true,
                    "nodes": [
                        {
                            "id": {
                                "py/id": 1
                            }
                        },
                        {
                            "id": {
                                "py/id": 2
                            }
                        },
                        {
                            "id": {
                                "py/id": 4
                            }
                        },
                        {
                            "id": {
                                "py/id": 5
                            }
                        },
                        {
                            "id": {
                                "py/id": 6
                            }
                        },
                        {
                            "id": {
                                "py/id": 7
                            }
                        },
                        {
                            "id": {
                                "py/id": 8
                            }
                        },
                        {
                            "id": {
                                "py/id": 9
                            }
                        },
                        {
                            "id": {
                                "py/id": 10
                            }
                        },
                        {
                            "id": {
                                "py/id": 11
                            }
                        },
                        {
                            "id": {
                                "py/id": 12
                            }
                        }
                    ]
                }
            ]
        },
        "preprocessing": {
            "ensure_rgb": false,
            "ensure_grayscale": true,
            "imagenet_mode": null,
            "input_scaling": 1.0,
            "pad_to_stride": 16,
            "resize_and_pad_to_target": true,
            "target_height": 482,
            "target_width": 559
        },
        "instance_cropping": {
            "center_on_part": null,
            "crop_size": null,
            "crop_size_detection_padding": 16
        }
    },
    "model": {
        "backbone": {
            "leap": null,
            "unet": {
                "stem_stride": null,
                "max_stride": 16,
                "output_stride": 2,
                "filters": 16,
                "filters_rate": 2.0,
                "middle_block": true,
                "up_interpolate": true,
                "stacks": 1
            },
            "hourglass": null,
            "resnet": null,
            "pretrained_encoder": null
        },
        "heads": {
            "single_instance": {
                "part_names": [
                    "L1",
                    "L2",
                    "L3",
                    "L4",
                    "R1",
                    "R2",
                    "R3",
                    "R4",
                    "Head",
                    "Middle",
                    "End"
                ],
                "sigma": 2.5,
                "output_stride": 2,
                "loss_weight": 1.0,
                "offset_refinement": false
            },
            "centroid": null,
            "centered_instance": null,
            "multi_instance": null,
            "multi_class_bottomup": null,
            "multi_class_topdown": null
        },
        "base_checkpoint": null
    },
    "optimization": {
        "preload_data": true,
        "augmentation_config": {
            "rotate": true,
            "rotation_min_angle": -15.0,
            "rotation_max_angle": 15.0,
            "translate": false,
            "translate_min": -5,
            "translate_max": 5,
            "scale": false,
            "scale_min": 0.9,
            "scale_max": 1.1,
            "uniform_noise": false,
            "uniform_noise_min_val": 0.0,
            "uniform_noise_max_val": 10.0,
            "gaussian_noise": false,
            "gaussian_noise_mean": 5.0,
            "gaussian_noise_stddev": 1.0,
            "contrast": false,
            "contrast_min_gamma": 0.5,
            "contrast_max_gamma": 2.0,
            "brightness": false,
            "brightness_min_val": 0.0,
            "brightness_max_val": 10.0,
            "random_crop": false,
            "random_crop_height": 256,
            "random_crop_width": 256,
            "random_flip": true,
            "flip_horizontal": false
        },
        "online_shuffling": true,
        "shuffle_buffer_size": 128,
        "prefetch": true,
        "batch_size": 4,
        "batches_per_epoch": 200,
        "min_batches_per_epoch": 200,
        "val_batches_per_epoch": 10,
        "min_val_batches_per_epoch": 10,
        "epochs": 200,
        "optimizer": "adam",
        "initial_learning_rate": 0.0001,
        "learning_rate_schedule": {
            "reduce_on_plateau": true,
            "reduction_factor": 0.5,
            "plateau_min_delta": 1e-06,
            "plateau_patience": 5,
            "plateau_cooldown": 3,
            "min_learning_rate": 1e-08
        },
        "hard_keypoint_mining": {
            "online_mining": false,
            "hard_to_easy_ratio": 2.0,
            "min_hard_keypoints": 2,
            "max_hard_keypoints": null,
            "loss_scale": 5.0
        },
        "early_stopping": {
            "stop_training_on_plateau": true,
            "plateau_min_delta": 1e-08,
            "plateau_patience": 10
        }
    },
    "outputs": {
        "save_outputs": true,
        "run_name": "230703_123601",
        "run_name_prefix": "June22_T3",
        "run_name_suffix": ".single_instance",
        "runs_folder": "models",
        "tags": [
            ""
        ],
        "save_visualizations": true,
        "delete_viz_images": true,
        "zip_outputs": false,
        "log_to_csv": true,
        "checkpointing": {
            "initial_model": false,
            "best_model": true,
            "every_epoch": false,
            "latest_model": false,
            "final_model": false
        },
        "tensorboard": {
            "write_logs": false,
            "loss_frequency": "epoch",
            "architecture_graph": false,
            "profile_graph": false,
            "visualizations": true
        },
        "zmq": {
            "subscribe_to_controller": false,
            "controller_address": "tcp://127.0.0.1:9000",
            "controller_polling_timeout": 10,
            "publish_updates": false,
            "publish_address": "tcp://127.0.0.1:9001"
        }
    },
    "name": "",
    "description": "",
    "sleap_version": "1.3.1",
    "filename": "models/June22_T3230703_123601.single_instance/training_config.json"
}
```

And of training_log.csv:
epoch | loss | lr | val_loss
-- | -- | -- | --
0 | 0.0002836266649 | 1.00E-04 | 0.0002806489938
1 | 0.0002782347728 | 1.00E-04 | 0.0002748701081
2 | 0.000265032053 | 1.00E-04 | 0.0002541888098
3 | 0.000250175508 | 1.00E-04 | 0.0002430039021
4 | 0.0002365944383 | 1.00E-04 | 0.000225908123
5 | 0.0002186617639 | 1.00E-04 | 0.0002135385002
6 | 0.0002016193757 | 1.00E-04 | 0.0001901303622
7 | 0.0001832203561 | 1.00E-04 | 0.0001711004006
8 | 0.000167876904 | 1.00E-04 | 0.000160680429
9 | 0.000158944109 | 1.00E-04 | 0.0001551544556
10 | 0.0001522106904 | 1.00E-04 | 0.0001490745199
11 | 0.0001460470667 | 1.00E-04 | 0.000140680786
12 | 0.0001401059853 | 1.00E-04 | 0.0001342108735
13 | 0.0001324612676 | 1.00E-04 | 0.0001361382892
14 | 0.0001264310558 | 1.00E-04 | 0.0001282223238
15 | 0.0001200535917 | 1.00E-04 | 0.000117416901
16 | 0.0001107626958 | 1.00E-04 | 0.0001070762301
17 | 0.0001030189669 | 1.00E-04 | 0.000101298865
18 | 9.51E-05 | 1.00E-04 | 9.42E-05
19 | 8.82E-05 | 1.00E-04 | 8.68E-05
20 | 8.31E-05 | 1.00E-04 | 8.43E-05
21 | 7.73E-05 | 1.00E-04 | 7.49E-05
22 | 7.35E-05 | 1.00E-04 | 7.46E-05
23 | 6.98E-05 | 1.00E-04 | 6.99E-05
24 | 6.66E-05 | 1.00E-04 | 6.74E-05

Thanks again,
Adi

**Comment by tacytown** [comment, marked as answer]
Hello again,

We figured out the problem! I was manually stopping the training when validation loss reached a point I was satisfied with, instead of letting it run to completion. When I let it run its course the .npz files generate just fine.

Thank you for the help!

Adi
